src/plugins/dashboard/public/application/lib/sync_dashboard_container_input.ts

```diff
@@ -0,0 +1,198 @@
+    applyFilters(latestState.query, input.filters);
+  }
+
+  if (!_.isEqual(input.panels, latestState.panels)) {
+    dispatchDashboardStateChange(setPanels(input.panels));
+  }
+
+  if (!_.isEqual(input.query, latestState.query)) {
+    dispatchDashboardStateChange(setQuery(input.query));
+  }
+
+  if (!_.isEqual(input.expandedPanelId, latestState.expandedPanelId)) {
+    dispatchDashboardStateChange(setExpandedPanelId(input.expandedPanelId));
+  }
+  dispatchDashboardStateChange(setFullScreenMode(input.isFullScreenMode));
+};
+
+export const applyStateChangesToContainer = ({
+  force,
+  history,
+  services,
+  savedDashboard,
+  dashboardContainer,
+  kbnUrlStateStorage,
+  isEmbeddedExternally,
+  getLatestDashboardState,
+}: ApplyStateChangesToContainerProps) => {
+  const latestState = getLatestDashboardState();
+  if (Object.keys(latestState).length === 0) {
+    return;
+  }
+  const currentDashboardStateAsInput = stateToDashboardContainerInput({
+    dashboardState: latestState,
+    isEmbeddedExternally,
+    savedDashboard,
+    services,
+  });
+  const differences = diffDashboardContainerInput(
+    dashboardContainer.getInput(),
+    currentDashboardStateAsInput
+  );
+  if (force) {
+    differences.lastReloadRequestTime = Date.now();
+  }
+
+  if (Object.keys(differences).length !== 0) {
+    const shouldRefetch = Object.keys(differences).some(
+      (changeKey) => !noRefetchKeys.includes(changeKey as keyof DashboardContainerInput)
+    );
+
+    const newSearchSessionId: string | undefined = (() => {
+      // do not update session id if this is irrelevant state change to prevent excessive searches
+      if (!shouldRefetch) return;
+
+      const sessionApi = services.data.search.session;
+      let searchSessionIdFromURL = getSearchSessionIdFromURL(history);
+      if (searchSessionIdFromURL) {
+        if (sessionApi.isRestore() && sessionApi.isCurrentSession(searchSessionIdFromURL)) {
+          // navigating away from a restored session
+          kbnUrlStateStorage.kbnUrlControls.updateAsync((nextUrl) => {
+            if (nextUrl.includes(DashboardConstants.SEARCH_SESSION_ID)) {
+              return replaceUrlHashQuery(nextUrl, (query) => {
+                delete query[DashboardConstants.SEARCH_SESSION_ID];
+                return query;
+              });
+            }
+            return nextUrl;
+          });
+          searchSessionIdFromURL = undefined;
+        } else {
+          sessionApi.restore(searchSessionIdFromURL);
+        }
+      }
+
+      return searchSessionIdFromURL ?? sessionApi.start();
+    })();
+
+    dashboardContainer.updateInput({
+      ...differences,
+      ...(newSearchSessionId && { searchSessionId: newSearchSessionId }),
+    });
+  }
+};
+
+const noRefetchKeys: Array<keyof DashboardContainerInput> = [
+  'title',
+  'viewMode',
+  'useMargins',
+  'description',
+  'expandedPanelId',
+  'isFullScreenMode',
+  'isEmbeddedExternally',
+];
```

Comment:
Pro tip: add a const assertion here.

https://blog.logrocket.com/const-assertions-are-the-killer-new-typescript-feature-b73451f35802/

Reply:
Very cool!